### ISSUE TITLE
Underreact Design

### ISSUE THREAD

**Issue by kepta** [mapbox/underreact#3]
TLDR; This post discusses the design decisions of Underreact

## What is Underreact?
Underreact will be a simple tool to build SPA application across Mapbox. The major goals of this project would be:
- Simple and performant
- Fits our need for small and large applications
- Is configurable but also works without configuration

## Why Underreact and not CRA?
I think it boils down to - **SPA applications are too important for us**. CRA tries to do a lot of things and not everything it does fits our use case (for eg. SCASS). It also doesn't let one customize the application and ejecting is pretty intimidating to think of. CRA is a great tool and a lot of the design decisions taken here are inspired by it. So if you love CRA, you would love Underreact too.

### HTML
HTML is pretty important, even for single page applications. Underreact will let users provide a function which would be expected to return an HTML string. This is similar to the HTML templating model found in Webpack, but a bit more powerful. The expected usage would be to set up a root element for frameworks like React and add any scripts/links if needed.
```javascript
// undereact.config.js
{
  htmlSource: ({ renderJsBundles, renderCssLinks }) => `
      <html>
      <head>
       <title>Words that rhyme with fish</title>
        <link href="https://api.mapbox.com/mapbox-assembly/v0.21.2/assembly.min.css" rel="stylesheet">
        ${renderCssLinks()}
      </head>
      <body>
        <div id="app">
          <!-- React app will be rendered into this div -->
        </div>
        ${renderJsBundles()}
      </body>
  
      </html>
    `;
}
```

### CSS
We would not allow the user to import CSS from within javascript which means no `import './index.css'`. We believe importing CSS is wrong since the import doesn't really give you anything tangible that can be used in the code. 

To allow the user to add stylesheets to their project, they can simply add paths to stylesheets in the configuration file. These files would then be processed and concatanated by PostCSS. By default Underreact would be using `Autoprefixer` and `CSSO` PostCSS plugins, but the user can also supply their own list of PostCSS plugins. 

Instead of using [webpack-dev-server](https://webpack.js.org/configuration/dev-server/), we have decided to go with [live-server](https://www.npmjs.com/package/live-server) which is much simpler and provides hot reloading of CSS. 

### Webpack
Webpack is currently the most popular tools out there for packaging Javascript applications. We do find it overly complex but the ecosystem and the popularity are hard to ignore.  Keeping Webpack configuration from spiraling, we have decided to keep CSS and HTML build steps independent from it. This makes Underreact less reliant on a single tool and will ameliorate future maintenance headaches.

### Importing Assets
Importing of assets with the ES module syntax is quite prevalent these days and not supporting it could hurt the adoption of Underreact. Hence, it would be support importing of JSON and images.

### Babel
Underreact would come with our own `babel-preset-mapbox`. This preset would serve as an abstraction to the common babel presets and plugins we use everywhere. It would also let the user override any of them. 

To keep onboarding simple, the `.babelrc` file is not required and a default Babel configuration is used behind the scenes. However, if the user wishes to customize Babel they are free to create their own `.babelrc`, which would supposedly use the `babel-preset-mapbox`. 

We found that customizing [browserslist](https://github.com/browserslist/browserslist) is pretty common and having a special option in `babel-preset-mapbox` for just this purpose would be super helpful. This option would also allow PostCSS to consume `browerslist`.

### underreact.config.json
  - `stylesheets`: A array contains the path to stylesheets the user wants to include in the bundle. 
  - `vendorModules`: Allows the user to keep the vendor bundle relatively stable. 
  - `devServerHistoryFallback`: hash mode by default, can enable the HTML5 browser history mode. 
  - `postcssPlugins`
  - `webpackLoaders`
  - `webpackPlugins`
  - `webpackConfigTransform`: A callback which can be used to mutate the webpack configuration.
  - `publicDirectory`: Always need some place to put files and just have them copied over.
  - `siteBasePath`:  The hostname of the web application, defaults to `/`.
  - `publicAssetsPath`: The path to place all the assets in, defaults to `/assets`.

### Other configurations
- We would be reading environment variables from a `.env` file, read more [here](https://github.com/motdotla/dotenv).
- Underreact would allow tweaking babel with the help of babelrc
- Best enjoyed with [stickler](https://github.com/mapbox/stickler)

@mapbox/frontend-platform @mapbox/frontend happy to hear your thoughts :)

**Comment by davidtheclark** [comment]
Awesome write-up, @kepta! Just two follow-up questions for you:

- Do you think it's necessary to pass `renderJsBundles` and `renderCssLinks` functions into the HTML-rendering function, or do you think we'd serve our use cases by automatically injecting those things (at the end of `</body>` and `</head>`, respectively)?
- What do you think about the HTML-rendering function returning either a string or a Promise that resolves with a string?

**Comment by davidtheclark** [comment]
@kepta What do you think our polyfill strategy should be?

**Comment by kepta** [comment]
> Do you think it's necessary to pass renderJsBundles and renderCssLinks functions into the HTML-rendering function, or do you think we'd serve our use cases by automatically injecting those things (at the end of </body> and </head>, respectively)?

I don't see any benefit in auto injecting them behind the scenes. On a plus side, pro-users would be happy to take command over the placement. 

> What do you think about the HTML-rendering function returning either a string or a Promise that resolves with a string?

Oops, I forgot to mention that, thanks for reminding. _(updated the ticket)_

**Comment by kepta** [comment]
> @kepta What do you think our polyfill strategy should be?

@davidtheclark getting inspired by [webpack](https://webpack.js.org/guides/shimming/), I propose that we use a default polyfill file and provide a `polyfillInject()` function in our async HTML-rednering function:
```
{
  htmlSource: ({ renderJsBundles, polyfillInject , renderCssLinks }) => `
      <html>
      <head>
       <title>Words that rhyme with fish</title>
         ${polyfillInject()}
        <link href="https://api.mapbox.com/mapbox-assembly/v0.21.2/assembly.min.css" rel="stylesheet">
        ${renderCssLinks()}
      </head>
      <body>
        <div id="app">
          <!-- React app will be rendered into this div -->
        </div>
        ${renderJsBundles()}
      </body>
      </html>
    `;
}
```
This `polyfillInject` function would do something similar to what webpack proposes, that is 
```
  var modernBrowser = (
    'fetch' in window &&
    'assign' in Object
  );

  if ( !modernBrowser ) {
    var scriptElement = document.createElement('script');

    scriptElement.async = false;
    scriptElement.src = '/polyfills.bundle.js';
    document.head.appendChild(scriptElement);
  }
```
Thoughts?

**Comment by davidtheclark** [comment]
> I don't see any benefit in auto injecting them behind the scenes.

The benefit would be that if it's always done in exactly the same way, you don't have to repeat the boilerplate every time. I wasn't sure about those functions any more because I couldn't think of any use cases for doing anything different with them: in every case, you should put the links at the end of `</head>` and the scripts at the end of `</body>`. Can you think of use cases for using those functions in any other way?

Interesting idea for polyfilling. So we'll need to decide which polyfills to include and what `modernBrowser` test covers them all sufficiently.

**Comment by kepta** [comment]
> Can you think of use cases for using those functions in any other way?

I think the functions provide a transparent way of templating the `html`. But if we are never gonna change their order we can possibly do them behind the scene.

> So we'll need to decide which polyfills to include and what modernBrowser test covers them all sufficiently.

Ya we will need a default list of polyfills, I wonder if the `browserslist` can help us with that?

**Comment by kepta** [comment]
Underreact was released 🎉 !
